File: scripts/check-report-bindings.py
#!/usr/bin/env python3
"""RPT-01 boundary test: every Column/Measure reference in a PBIR report must resolve to a VISIBLE
object in the semantic model's TMDL — at the reviewed commit when one is given.

Usage: check-report-bindings.py <Report>.Report <Model>.SemanticModel [--at <git-sha>]
  --at  read the model's TMDL from that commit (git show) instead of the working tree.
        Pass the `model_commit` from the latest passing model review.
Exit 1 on findings."""
import sys, json, pathlib, re, subprocess

args = sys.argv[1:]
rep = pathlib.Path(args[0]); model = pathlib.Path(args[1])
at = args[args.index("--at") + 1] if "--at" in args else None

def tmdl_files():
    tables = model / "definition" / "tables"
    if not at:
        return [(f.name, f.read_text(encoding="utf-8")) for f in sorted(tables.glob("*.tmdl"))]
    top = subprocess.run(["git", "rev-parse", "--show-toplevel"], capture_output=True, text=True, check=True).stdout.strip()
    rel = str(tables.resolve().relative_to(top))
    names = subprocess.run(["git", "ls-tree", "--name-only", at, rel + "/"], capture_output=True, text=True, check=True).stdout.split()
    return [(n.rsplit("/", 1)[-1], subprocess.run(["git", "show", f"{at}:{n}"], capture_output=True, text=True, check=True).stdout) for n in names if n.endswith(".tmdl")]

cols, meas = set(), set()
for _, txt in tmdl_files():
    m = re.search(r"^table\s+(?:'([^']+)'|(\S+))", txt, re.M); table = m.group(1) or m.group(2)
    head = txt.split("\n\tcolumn", 1)[0].split("\n\tmeasure", 1)[0]
    table_hidden = re.search(r"^\tisHidden\s*$", head, re.M) is not None
    for c in re.finditer(r"^\tcolumn\s+(?:'([^']+)'|(\S+))(.*?)(?=^\t(?:column|measure|partition|hierarchy)\b|\Z)", txt, re.M | re.S):
        if "isHidden" not in c.group(3) and not table_hidden: cols.add((table, c.group(1) or c.group(2)))
    for mm in re.finditer(r"^\tmeasure\s+(?:'([^']+)'|([^=\s]+))\s*=(.*?)(?=^\t(?:column|measure|partition|hierarchy)\b|^\t///|\Z)", txt, re.M | re.S):
        if "isHidden" not in mm.group(3): meas.add((table, mm.group(1) or mm.group(2)))

findings, refs = [], 0
def walk(o, where):
    global refs
    if isinstance(o, dict):
        for key, allowed in (("Column", cols), ("Measure", meas)):
            if key in o and isinstance(o[key], dict) and "Property" in o[key]:
                ent = o[key].get("Expression", {}).get("SourceRef", {}).get("Entity", "?"); prop = o[key]["Property"]; refs += 1
                if (ent, prop) not in allowed:
                    findings.append(f"{where}: {key.lower()} '{ent}'[{prop}] is not a visible object in the model{' at ' + at[:7] if at else ''} (RPT-01)")
        for v in o.values(): walk(v, where)
    elif isinstance(o, list):
        for v in o: walk(v, where)

for f in sorted(rep.rglob("*.json")):
    if f.name in ("version.json", "report.json", "pages.json"): continue
    try: data = json.loads(f.read_text(encoding="utf-8"))
    except Exception as e: findings.append(f"{f}: invalid JSON ({e})"); continue
    walk(data, str(f.relative_to(rep)))
    if f.name == "visual.json" and ("measures" in data.get("visual", {}) or "reportMeasures" in data):
        findings.append(f"{f.relative_to(rep)}: report-level measure defined (RPT-02)")
for x in findings: print("[bindings] " + x)
print(f"[bindings] {rep.name} vs {model.name}{' @ ' + at[:7] if at else ' (working tree)'}: {refs} references, {len(findings)} findings")
sys.exit(1 if findings else 0)

File: scripts/check-report-bindings.js
#!/usr/bin/env node
// RPT-01 boundary test: every Column/Measure reference in a PBIR report must resolve to a VISIBLE
// object in the model's TMDL — at the reviewed commit when --at <sha> is given.
//   usage: node check-report-bindings.js <Report>.Report <Model>.SemanticModel [--at <sha>]
"use strict";
const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");
const { walk, norm } = require("./lib.js");

const args = process.argv.slice(2);
const rep = args[0], model = args[1];
const at = args.includes("--at") ? args[args.indexOf("--at") + 1] : null;

function tmdlFiles() {
  const tables = path.join(model, "definition", "tables");
  if (!at) return fs.readdirSync(tables).filter((n) => n.endsWith(".tmdl")).sort().map((n) => fs.readFileSync(path.join(tables, n), "utf8"));
  const top = execSync("git rev-parse --show-toplevel", { cwd: model, encoding: "utf8" }).trim();
  const rel = norm(path.relative(top, path.resolve(tables)));
  const names = execSync(`git ls-tree --name-only ${at} "${rel}/"`, { cwd: top, encoding: "utf8" }).split(/\r?\n/).filter((n) => n.endsWith(".tmdl"));
  return names.map((n) => execSync(`git show ${at}:"${n}"`, { cwd: top, encoding: "utf8" }));
}
const cols = new Set(), meas = new Set();
for (const txt of tmdlFiles()) {
  const tm = txt.match(/^table\s+(?:'([^']+)'|(\S+))/m); const table = tm[1] || tm[2];
  const head = txt.split(/\n\tcolumn|\n\tmeasure/)[0];
  const tableHidden = /^\tisHidden\s*$/m.test(head);
  for (const c of txt.matchAll(/^\tcolumn\s+(?:'([^']+)'|(\S+))([\s\S]*?)(?=^\t(?:column|measure|partition|hierarchy)\b|(?![\s\S]))/gm))
    if (!/isHidden/.test(c[3]) && !tableHidden) cols.add(`${table}|${c[1] || c[2]}`);
  for (const m of txt.matchAll(/^\tmeasure\s+(?:'([^']+)'|([^=\s]+))\s*=([\s\S]*?)(?=^\t(?:column|measure|partition|hierarchy)\b|^\t\/\/\/|(?![\s\S]))/gm))
    if (!/isHidden/.test(m[3])) meas.add(`${table}|${m[1] || m[2]}`);
}
const findings = []; let refs = 0;
function visit(o, where) {
  if (Array.isArray(o)) return o.forEach((v) => visit(v, where));
  if (!o || typeof o !== "object") return;
  for (const [key, set] of [["Column", cols], ["Measure", meas]]) {
    const x = o[key];
    if (x && typeof x === "object" && "Property" in x) {
      const ent = (((x.Expression || {}).SourceRef) || {}).Entity || "?"; refs++;
      if (!set.has(`${ent}|${x.Property}`)) findings.push(`${where}: ${key.toLowerCase()} '${ent}'[${x.Property}] is not a visible object in the model${at ? " at " + at.slice(0, 7) : ""} (RPT-01)`);
    }
  }
  for (const v of Object.values(o)) visit(v, where);
}
for (const f of walk(rep, (p) => p.endsWith(".json") && !/(version|report|pages)\.json$/.test(p)).sort()) {
  const where = norm(path.relative(rep, f)); let data;
  try { data = JSON.parse(fs.readFileSync(f, "utf8")); } catch (e) { findings.push(`${where}: invalid JSON (${e.message})`); continue; }
  visit(data, where);
  if (f.endsWith("visual.json") && ((data.visual || {}).measures || data.reportMeasures)) findings.push(`${where}: report-level measure defined (RPT-02)`);
}
for (const x of findings) console.log("[bindings] " + x);
console.log(`[bindings] ${path.basename(rep)} vs ${path.basename(model)}${at ? " @ " + at.slice(0, 7) : " (working tree)"}: ${refs} references, ${findings.length} findings`);
process.exit(findings.length ? 1 : 0);
